Rewrite match subscreen to remove bindables

smoogipoo opened the pull request:
Fixes https://github.com/ppy/osu/issues/32503
Split out from https://github.com/ppy/osu/pull/32250, with minor changes.

This finalises the refactorings that have been going on for a while now.

In a similar fashion to #31882, this is a full rewrite of `MultiplayerMatchSubScreen` that no longer inherits (and finally removes) `RoomSubScreen`. The primary goal here is to remove the beatmap/ruleset/mod/"selected item" bindables and use `MultiplayerRoom` states.

Because it's a rewrite, I suggest going through the screen anew (rather than looking at the diff) and possibly comparing to the implementation of `PlaylistsRoomSubScreen`.

smoogipoo committed:
Rewrite match subscreen to use full online state

smoogipoo committed:
Remove unnecessary using

smoogipoo committed:
Merge branch 'master' into refactor-match-subscreen

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs:
```diff
@@ -28,410 +35,869 @@
+        /// Responds to changes in the active room to adjust the visibility of the settings and main content.
+        /// Only the settings overlay is visible while the room isn't created, and only the main content is visible after creation.
+        /// </summary>
+        private void onRoomUpdated() => Scheduler.AddOnce(() =>
```
`Scheduler.AddOnce()` does not debounce correctly if it's given an inline lambda like this because it's going to fail an equality check framework-side to determine if it should debounce or not. The two options are to either make the body of this a named method (this is the most idiomatic one we use), or to utilise the closure-eliminating overload of this, and have the inner lambda be a `static (MultiplayerMatchSubScreen screen) => { }` (don't recommend that one).

smoogipoo replied:
I don't get it. Have you actually tested this? This is very idiomatic code in multiplayer, and it would also mean stuff like this is broken:

https://github.com/ppy/osu/blob/3852cca19aa873ed9ba69dc7bef88580d8573440/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerRoomSounds.cs#L42-L55

Which was added _not_ by me, here: https://github.com/ppy/osu/commit/ef5e37c47a1099c393238d53e8e80f9621c1b06a. If this is the case I will do everything in my power to obliterate this method.

bdach replied:
No, I have not tested this specific code with a debugger. I did not remember the test above existed. I wrote something like

```csharp
        [Test]
        public void TestAddOnce()
        {
            var scheduler = new Scheduler();
            int invocations = 0;

            scheduler.AddOnce(() => invocations++);
            scheduler.AddOnce(() => invocations++);
            scheduler.Update();

            Assert.That(invocations, Is.EqualTo(1));
        }
```

and *that* fails. Which I guess isn't strictly the same what you have because you're never repeating that lambda again, *but* as that code comment says, I've at this point internalised that `AddOnce()` should not be called on inline lambdas because there be dragons and it might or it might not work as intended.

smoogipoo replied:
I believe the way it's conceptualised is that the lambda instance itself is always the same. i.e.

```
static object? a;

void m()
{
    a ??= new object();
}

static object? b;
static object? c;

void n()
{
    b ??= new object();
    c ??= new object();
}
```

The usage in multiplayer is always a case of the former.

bdach replied:
> The usage in multiplayer is always a case of the former.

That's not even it actually. If you pull up the IL display in rider and put it to low level you'll see this:

```csharp
    private void onRoomUpdated()
    {
      this.Scheduler.AddOnce(new Action((object) this, __methodptr(<onRoomUpdated>b__77_0)));
    }
```

This is *allocating a delegate instance every time*, but the compiler has *figured out that it can make a method group delegate out of the body of the lambda*, so the delegate equality check inside scheduler doesn't fail (because the lambdas are attached to a named method). This'll only work as long as the compiler does *that* and nothing else.

smoogipoo replied:
Are you fine with it as is then? As I said this is pretty standard through all of multiplayer, feels weird if we're applying any changes only to this usage.

bdach replied:
I suppose we can keep it until a problem sure.

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs:
```diff
@@ -28,410 +35,869 @@
+                return base.OnBackButton();
             }
 
-            // The beatmap is queried asynchronously when the selected item changes.
-            // This is an issue with MultiSpectatorScreen which is effectively in an always "ready" state and receives LoadRequested() callbacks
-            // even when it is not truly ready (i.e. the beatmap hasn't been selected by the client yet). For the time being, a simple fix to this is to ignore the callback.
-            // Note that spectator will be entered automatically when the client is capable of doing so via beatmap availability callbacks (see: updateBeatmapAvailability()).
-            if (client.LocalUser?.State == MultiplayerUserState.Spectating && (SelectedItem.Value == null || Beatmap.IsDefault))
```
Is whatever happened here no longer relevant? I don't see this logic moved anywhere.

(I'd honestly take a "can be ignored" in a heartbeat because I'm not even sure what this inline comment is trying to say. Blame points at https://github.com/ppy/osu/pull/16773.)

smoogipoo replied:
Right, the history of this was lost in the former PR. Although even there it's a little confusing.

Basically, this is expected to be handled by the online availability tracker and fail on [this check](https://github.com/ppy/osu/blob/f288cfc4648bc922c671c0c75dc3b22d4b84177d/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs#L521-L522). Somewhat particularly important to this is that the tracker [does not](https://github.com/ppy/osu/blob/2dc38b5f094746f2933852f812458863719e5307/osu.Game/Screens/OnlinePlay/OnlinePlayBeatmapAvailabilityTracker.cs#L58-L72) [schedule](https://github.com/ppy/osu/blob/2dc38b5f094746f2933852f812458863719e5307/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerBeatmapAvailabilityTracker.cs#L24-L30) its update, so it will always initially be `Unknown` when the current item changes.

On its own, I can see why that is quite worrying, and it would make sense to ensure we're in the correct state here. I'll look at adding that.

smoogipoo replied:
Hopefully this makes more sense with https://github.com/ppy/osu/pull/32669/commits/71af50f67541fcac80c4e1e97b8028aad761a3c3 though the above comment is no longer applicable (either here or `master`).

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs:
```diff
@@ -28,410 +35,869 @@
+
+            // Update global gameplay state to correspond to the new selection.
+            // Retrieve the corresponding local beatmap, since we can't directly use the playlist's beatmap info
+            var localBeatmap = beatmapManager.QueryBeatmap(b => b.OnlineID == gameplayBeatmapId);
```
Hash check added in https://github.com/ppy/osu/pull/32596 has been removed here

bdach reviewed (changes requested):
Well this one didn't get much easier to review.

Spotted a few things from reading the code, once those are fixed I'll probably just do a test in-game and click buttons if I can't break it.

smoogipoo committed:
Add back removed hash check

smoogipoo committed:
Validate state with lesser magic

bdach reviewed (changes requested):
Empirically tested this some. Found only one thing, namely that clicking the countdown button on the bottom bar crashes with

```
[runtime] 2025-04-08 08:42:03 [verbose]: Too many unhandled exceptions, crashing out.
Unhandled exception. System.InvalidOperationException: Cannot show or hide a popover without a parent PopoverContainer in the hierarchy
   at osu.Framework.Extensions.PopoverExtensions.setTargetOnNearestPopover(Drawable origin, IHasPopover target)
   at osu.Framework.Extensions.PopoverExtensions.ShowPopover(IHasPopover hasPopover)
   at osu.Framework.Graphics.Containers.ClickableContainer.OnClick(ClickEvent e)
   at osu.Game.Graphics.UserInterface.OsuAnimatedButton.OnClick(ClickEvent e) in /Users/bdach/Documents/Work/open-source/osu/osu.Game/Graphics/UserInterface/OsuAnimatedButton.cs:line 117
   at osu.Framework.Input.ButtonEventManager`1.PropagateButtonEvent(IEnumerable`1 drawables, UIEvent e)
   at osu.Framework.Input.MouseButtonEventManager.handleClick(InputState state, List`1 targets)
   at osu.Framework.Input.MouseButtonEventManager.HandleButtonUp(InputState state, List`1 targets)
```

smoogipoo committed:
Add failing countdown start test

smoogipoo committed:
Fix crash when trying to start countdown

smoogipoo commented:
Good catch, didn't have any tests for that.

smoogipoo committed:
Clone countdown objects in test client

Beyond isolation, this stuff is also used to test (de-)serialisation.
Much-of-a-muchness in this case since it isn't too thoroughly tested,
but best do it anyway.
